Firebase.cpp

```diff
@@ -15,98 +15,147 @@
 //
 #include "Firebase.h"
 
-const char* firebaseFingerprint = "7A 54 06 9B DC 7A 25 B3 86 8D 66 53 48 2C 0B 96 42 C7 B3 0A";
-const uint16_t firebasePort = 443;
+namespace {
+const char* kFirebaseFingerprint = "7A 54 06 9B DC 7A 25 B3 86 8D 66 53 48 2C 0B 96 42 C7 B3 0A";
+const uint16_t kFirebasePort = 443;
 
-Firebase::Firebase(const String& host) : _host(host) {
-  _http.setReuse(true);
+String makeFirebaseURL(const String& path, const String& auth) {
+  String url;
+  if (path[0] != '/') {
+    url = "/";
+  }
+  url += path + ".json";
+  if (auth.length() > 0) {
+    url += "?auth=" + auth;
+  }
+  return url;
+}
+
+}  // namespace
+
+Firebase::Firebase(const String& host) : host_(host) {
+  http_.setReuse(true);
 }
 
 Firebase& Firebase::auth(const String& auth) {
-  _auth = auth;
+  auth_ = auth;
   return *this;
 }
 
-String Firebase::get(const String& path) {
-  return sendRequest("GET", path);
+FirebaseGet Firebase::get(const String& path) {
+  return FirebaseGet(host_, auth_, path, &http_);
+}
+
+FirebasePush Firebase::push(const String& path, const String& value) {
+  return FirebasePush(host_, auth_, path, value, &http_);
+}
+
+FirebaseRemove Firebase::remove(const String& path) {
+  return FirebaseRemove(host_, auth_, path, &http_);
 }
 
-String Firebase::push(const String& path, const String& value) {
-  return sendRequest("POST", path, value);
+FirebaseStream Firebase::stream(const String& path) {
+  return FirebaseStream(host_, auth_, path);  // stream doesn't reuse http client.
 }
 
-Firebase& Firebase::stream(const String& path) {
-  _error.reset();
-  String url = makeURL(path); 
-  const char* headers[] = {"Location"};
-  _http.setReuse(true);  
-  _http.begin(_host.c_str(), firebasePort, url.c_str(), true, firebaseFingerprint);
-  _http.collectHeaders(headers, 1);
-  _http.addHeader("Accept", "text/event-stream");
-  int statusCode = _http.sendRequest("GET", (uint8_t*)NULL, 0);
-  String location;
-  // TODO(proppy): Add a max redirect check
-  while (statusCode == 307) {
-    location = _http.header("Location");
-    _http.setReuse(false);
-    _http.end();
-    _http.setReuse(true);    
-    _http.begin(location, firebaseFingerprint);
-    statusCode = _http.sendRequest("GET", (uint8_t*)NULL, 0);
+// FirebaseCall
+FirebaseCall::FirebaseCall(const String& host, const String& auth,
+			   const char* method, const String& path,
+			   const String& data, HTTPClient* http) {
+  if (!http) {
```

Comment:
We should probably reuse the client for now (even for stream) and do that in a separate PR (that might be the cause of the crash we see with the stream sample)